### DISCUSSION TITLE
lookup error using union()

### DISCUSSION THREAD

**Discussion by teafmartins** [serenity-is/Serenity#7168] [Help]
Hello,

What's wrong in this lookup using UNION() ?
Without using UNION it works fine!

```
public class MlLeagueFriendLookup2 : RowLookupScript<MlLeagueFriendRow>
{
    public ITwoLevelCache TwoLevelCache { get; }
    public IUserAccessor UserAccessor { get; }
    public MlLeagueFriendLookup2(ISqlConnections sqlConnections, ITwoLevelCache twoLevelCache, IUserAccessor userAccessor) : base(sqlConnections)
    {
        Expiration = TimeSpan.FromDays(-1);
        TwoLevelCache = twoLevelCache ?? throw new ArgumentNullException(nameof(twoLevelCache));
        UserAccessor = userAccessor ?? throw new ArgumentNullException(nameof(userAccessor));
    }
    protected override void PrepareQuery(SqlQuery query)
    {

        var fld = MlLeagueFriendRow.Fields;
        var UserId = Convert.ToInt32(UserAccessor.User.GetIdentifier());

        
        query.Distinct(true)
            .From("Users")
            .SelectAs("UserId", fld.LfFriendId).SelectAs("NickName", fld.FriendName)
            .Where(new Criteria("UserId") == UserId)
            .Union()
            .From("MLLeagueFriend")
            .Select(fld.LfFriendId)
            .Select(fld.FriendName)
            .Where(new Criteria(fld.LfUserId) == UserId);
    }

    protected override void ApplyOrder(SqlQuery query)
    {
    }
}

```

Thanks for your help

**Comment by LSNicholls** [comment]
What is the error you get?

**Reply by teafmartins** [reply]
@LSNicholls , when i click on quickfilter i get the error:
"An error occurred while trying to load lookup MLeagueFriendLookup2..
![image](https://github.com/serenity-is/Serenity/assets/17210232/e317ac33-f959-42c2-bd2c-36f722b1e9a9)
Can you help?

**Reply by VictorTomaili** [reply]
Did you check the console for more info?

**Comment by LSNicholls** [comment]
@teafmartins, Victor has told you the right thing to do to see the error.  In the console output you should see 2 things:

1. you should see the generated UNION statement.  Can you check that statement to make sure it is correct?  Can you run that statement interactively in a SQL IDE and ensure that it runs and provides the result you expect ?
2. you should see the actual error that occurs when Serenity runs it for the lookup in the console as well.  What is it?

**Comment by teafmartins** [comment]
@LSNicholls , @VictorTomaili , with your ideas I checked that the sql statement was invalid and changed the code to:

```
        query.Select(fld.LfFriendId)
            .Select(fld.FriendNickName)
            .Where(new Criteria(fld.LfUserId) == UserId)
            .Union()
            .From("Users")
            .SelectAs("UserId", fld.LfFriendId)
            .SelectAs("NickName", fld.FriendNickName)
            .Where(new Criteria("UserId") == UserId);
```

This sql statement is valid and it returns results, 
![image](https://github.com/serenity-is/Serenity/assets/17210232/06fcb931-aa8f-4115-8836-308d1cad2202)
No erros now, but the box appears with blank items, as follows:
![image](https://github.com/serenity-is/Serenity/assets/17210232/b521aa03-2da2-431e-ab66-a7543a552b1a)
Any idea why this happens?

Thank You

**Reply by LSNicholls** [reply]
@teafmartins, I don't know squat about that syntax; I'm a SQL person and if it were me I would just create a view that did what I want and use that as the entity behind the lookup.   :-)

But two possibilities occur to me 

1. is it possible that the lookup is unhappy because your rows are no longer distinct? (IOW if the key value appears more than once, I can see that being a problem)
2. more likely: it looks like your column names are no longer the same as they would have been earlier with your SelectAs clause.  So, if your FriendLookup is expecting a column named "NickName", now it doesn't have that column, and if expects a column named "UserId", now it doesn't have that column either.

**Comment by teafmartins** [comment]
@LSNicholls , I had already thought of the solution using a view... I'll try it this way.
Thanks for your time.

**Reply by LSNicholls** [reply]
Well, I still think the only problem with the union is probably your column names <g>.  And that problem could of course exist with a view, it's just easier to see!

**Comment by teafmartins** [comment]
Another solution is not to use union and add aditional quickfilter items when loading grid (grid constructor).
